ARROW-10889: [Rust] [Proposal] Add guidelines about usage of `unsafe`

jorgecarleitao opened the pull request:
I would like to propose that we outline and enforce guidelines on the arrow crate implementation with respect to the usage of `unsafe`.

The background of this proposal are PRs #8645 and #8829. In both cases, while addressing an unrelated issue, they hit undefined behavior (UB) due to an incorrect usage of `unsafe` in the code base. This UB was very time-consuming to identify and debug: combined, they accounted for more than 12hs of my time.

Safety against undefined behavior is the core premise of the Rust language. In many cases, the maintenance burden (time to find and fix bugs) does not justify the performance improvements and the decrease in motivation in handling them (they are just painful due to how difficult they are to debug). In particular, IMO those 12 hours would have been better spent in other parts of the code if `unsafe` would have not been used in the first place, which would have been likely translated in faster code or more features.

There are situations where `unsafe` is necessary, and the guidelines outline these cases. However, I also see many uses of `unsafe` that are not necessary nor properly documented.

The goal of these guidelines is to motivate contributors of the Rust implementation to be conscious about the maintenance cost of `unsafe`, and outline specific necessary conditions for any new `unsafe` to be introduced in the code base.

andygrove commented:
This is funny timing. I just filed ARROW-10889 for this and then saw this PR!

andygrove reviewed (approved):
This looks great. Thanks @jorgecarleitao !

andygrove commented:
It would be really cool to have a GitHub Action that detected `unsafe` in a PR and auto-posted a link to this documentation, as well as tagging the PR with an `unsafe` label. I do not know how feasible that is though.

Dandandan commented:
Very cool proposal. Will take some time soon to read it in detail. I think currently there is a lot of code violating what is proposed, so that's clearly needed 👍 

I think it also makes sense to keep track of the amount of `unsafe` that is used, and try to reduce it as much as possible by introducing safe APIs that are as performant. An example might be: introducing a safe iterator to avoid bounds checking. 

Also I think the unsafe code can /should be more tested, e.g. by having more `debug_asserts` etc. checking for invariants and checking for undefined behavior in the CI.

Dandandan commented on rust/arrow/README.md:
```diff
@@ -93,6 +93,79 @@ Arrow uses the following features:
+`[i * size_of<i32>(), (i + 1) * size_of<i32>()]`, and then transmuting it to `i32`. In safe Rust, 
+this operation requires boundary checks that are detrimental to performance.
+
+Usage of `unsafe` for performance reasons is justified only when the performance difference of a publicly available API is estatistically significantly larger than 10%, as demonstrated by a `bench`.
```
Maybe 10% is a bit arbitrary and depends on context? Maybe it should.bd demonstrated / be likely that it has an impact on real world usage of arrow?

alamb replied:
I agree that this wording would be stronger without a specific bound. I would rather say something like "usage of unsafe for performance reasons is justified only when all other alternatives have been exhausted and the performance benefits are sufficiently large. Performance benefits should be quantified with a `bench`".

jorgecarleitao replied:
I understand the sentiment.

My concern is that I am not sure I would like to maintain a complex `unsafe` code for a 6% improvement _at this particular phase_ of the project.

The rational for a concrete number is to impose a bound that we consider to not have the capacity to handle the maintenance cost for "small" improvements, and thus people should not try to focus on those types of improvements (again, _at this particular phase_ of the project).

My concern with "performance benefits are sufficiently large" is that anything bigger than zero is always "sufficiently large" when compared to zero.

What if we write something like

> usage of unsafe for performance reasons is justified only when all other alternatives have been exhausted and the performance benefits are sufficiently large (>~10%)

so that we allow other values, but we offer a number for reference?

ritchie46 replied:
Could the improvement be expressed by the **"hotness"** of the code? I can imagine that a 6% improvement in the `Buffer`s is much more valuable than a 11% percentage improvement in a specific kernel.

alamb replied:
I like

> usage of unsafe for performance reasons is justified only when all other alternatives have been exhausted and the performance benefits are sufficiently large (e.g. >~10%)

@ritchie46  I agree with the sentiments that some performance improvements are more important than others, which is why I was suggesting leaving room for interpretation in the writeup

Dandandan replied:
Yes, I think it would be preferable to have some room to trade safety for performance and vice versa. A change that makes everything a few percent faster might be acceptable, while a 50% improvement on a micro benchmark might not in some cases. And in some cases you maybe want to reduce the amount of safety, but might want to accept a certain negative performance impact.

Dandandan commented on rust/arrow/README.md:
```diff
@@ -93,6 +93,79 @@ Arrow uses the following features:
+* have code documentation for why a `safe` is not used / possible (e.g. `// 30% performance degradation if the safe counterpart is used, see bench X`)
+* have code documentation about which invariant the user needs to enforce to ensure no undefined behavior (e.g. `// this buffer must be constructed according to the arrow specification`)
+* if applicable, have the necessary `assert`s (not `debug_assert`!) of invariants.
+
```
I think places that do use unsafe, should have their invariants in a debug assert (as it's often something you want to skill for release builds).

alamb replied:
I also don't fully understand the mandate about `assert` -- I suggest removing the last bullet point as I think it may be more confusing than helpful

alamb commented on rust/arrow/README.md:
```diff
@@ -93,6 +93,79 @@ Arrow uses the following features:
+## Guidelines in usage of `unsafe`
+
+[`unsafe`](https://doc.rust-lang.org/book/ch19-01-unsafe-rust.html) has a high maintenance cost because debugging and testing it is difficult, time consuming, often requires external tools (e.g. `valgrind`), and requires a higher-than-usual attention to details. Undefined behavior is particularly difficult to identify and test, and usage of `unsafe` is the primary cause of undefined behavior in a ...
+
```
```suggestion
For two real world examples of where `unsafe` has consumed time in the past in this project see [#8545](https://github.com/apache/arrow/pull/8645) and [8829](https://github.com/apache/arrow/pull/8829)
```

alamb commented on rust/arrow/README.md:
```diff
@@ -93,6 +93,79 @@ Arrow uses the following features:
 
+## Guidelines in usage of `unsafe`
+
+[`unsafe`](https://doc.rust-lang.org/book/ch19-01-unsafe-rust.html) has a high maintenance cost because debugging and testing it is difficult, time consuming, often requires external tools (e.g. `valgrind`), and requires a higher-than-usual attention to details. Undefined behavior is particularly difficult to identify and test, and usage of `unsafe` is the primary cause of undefined behavior in a ...
```
https://doc.rust-lang.org/reference/behavior-considered-undefined.html might be a reasonable citation, though it doesn't quantify the sources of unsafetly.

alamb commented on rust/arrow/README.md:
```diff
@@ -93,6 +93,79 @@ Arrow uses the following features:
+
+### When can `unsafe` be used?
+
+Generally, `unsafe` can only be used when a `safe` counterpart is not available or has performance implications. The following is a summary of the current components of the crate that require `unsafe`:
+
```
```suggestion
Generally, `unsafe` should only be used when a `safe` counterpart is not available and there is no `safe` way to achieve additional performance in that area. The following is a summary of the current components of the crate that require `unsafe`:

```

alamb commented on rust/arrow/README.md:
```diff
@@ -93,6 +93,79 @@ Arrow uses the following features:
+* Foreign interfaces (C data interface)
+* Inter-process communication (IPC)
+* SIMD
+* Performance
```
I suggest being more explicit here -- maybe "Omitting bounds checking for performance" or something

alamb commented on rust/arrow/README.md:
```diff
@@ -93,6 +93,79 @@ Arrow uses the following features:
+The arrow format recommends storing buffers aligned with cache lines, and this crate adopts this behavior.
+However, Rust's global allocator does not allocates memory aligned with cache-lines. As such, many of the low-level operations related to memory management require `unsafe`.
+
+Usage of `unsafe` for the purposes of supporting allocations aligned with cache lines is allowed.
```
stylistically, I think we could remove the 'Usage of `unsafe` for ...' sentences. The rationale for use of unsafe is already explained above so this repetition seems redundant to me.

I also think such wording may give the impression that `unsafe` is always allowed for these operations, when I think the intent is "unsafe can be used as a last resort"

alamb commented on rust/arrow/README.md:
```diff
@@ -93,6 +93,79 @@ Arrow uses the following features:
+Usage of `unsafe` in this crate *must*:
+
+* not expose a public API as `safe` when there are necessary invariants for that API to be defined behavior.
+* have code documentation for why a `safe` is not used / possible (e.g. `// 30% performance degradation if the safe counterpart is used, see bench X`)
```
```suggestion
* have code documentation for why `safe` is not used / possible (e.g. `// 30% performance degradation if the safe counterpart is used, see bench X`)
```

alamb reviewed (approved):
I think this is a great idea -- thank you for writing it up @jorgecarleitao !

alamb commented:
@jorgecarleitao  I am not sure what has happened to this PR but it now seemingly includes many changes that are unrelated to the readme improvements:
<img width="1384" alt="Screen Shot 2020-12-14 at 1 24 21 PM" src="https://user-images.githubusercontent.com/490673/102119836-bc543980-3e0f-11eb-93dd-6a8d84c215ed.png">

jorgecarleitao commented:
Some mistake on my part. Fixed

alamb reviewed (approved):
I think this looks good. @sunchao  @paddyhoran or @nevi-me  would you like to express an opinion on this proposal?

Dandandan reviewed (approved):
For me the changes are clear enough and provide a good guideline for using unsafe.

andygrove commented:
I saw a tweet yesterday from @timClicks showing an example of how the imageproc project documents their unsafe code and I really like it. I think we should do something similiar. Maybe we could add this to the guidelines @jorgecarleitao ?

```rust
            // JUSTIFICATION
            //  Benefit
            //      Using checked indexing here makes bench_integral_image_rgb take 1.05x as long
            //      (The results are noisy, but this seems to be reproducible. I've not checked the generated assembly.)
            //  Correctness
            //      x and y are within bounds by definition of in_width and in_height
            let input = unsafe { image.unsafe_get_pixel(x, y) };
```

Dandandan commented:
> I saw a tweet yesterday from @timClicks showing an example of how the imageproc project documents their unsafe code and I really like it. I think we should do something similiar. Maybe we could add this to the guidelines @jorgecarleitao ?
> 
> ```rust
>             // JUSTIFICATION
>             //  Benefit
>             //      Using checked indexing here makes bench_integral_image_rgb take 1.05x as long
>             //      (The results are noisy, but this seems to be reproducible. I've not checked the generated assembly.)
>             //  Correctness
>             //      x and y are within bounds by definition of in_width and in_height
>             let input = unsafe { image.unsafe_get_pixel(x, y) };
> ```

Saw that too :) I think it's an interesting approach. In this PR https://github.com/apache/arrow/pull/8900 I added a comment with "SAFETY", but that misses the benefit part.

paddyhoran commented on rust/arrow/README.md:
```diff
@@ -96,6 +96,71 @@ Arrow uses the following features:
+#### cache-line aligned memory management
+
+The arrow format recommends storing buffers aligned with cache lines, and this crate adopts this behavior.
+However, Rust's global allocator does not allocates memory aligned with cache-lines. As such, many of the low-level operations related to memory management require `unsafe`.
```
```suggestion
However, Rust's global allocator does not allocate memory aligned with cache-lines. As such, many of the low-level operations related to memory management require `unsafe`.
```

paddyhoran commented on rust/arrow/README.md:
```diff
@@ -96,6 +96,71 @@ Arrow uses the following features:
+#### Interpreting bytes
+
+The arrow format is specified in bytes (`u8`), which can be logically represented as certain types
+depending on the DataType.
```
```suggestion
depending on the `DataType`.
```

paddyhoran commented on rust/arrow/README.md:
```diff
@@ -96,6 +96,71 @@ Arrow uses the following features:
+
+The arrow format declares an ABI for zero-copy from and to libraries that implement the specification
+(foreign interfaces). In Rust, receiving and sending pointers via FFI requires usage of `unsafe` due to 
+the impossibility of the compiler to derive the invariants (such as lifetime, null pointers, and pointer alignment), from the source code alone as they are part of the FFI contract.
```
```suggestion
the impossibility of the compiler to derive the invariants (such as lifetime, null pointers, and pointer alignment) from the source code alone as they are part of the FFI contract.
```

paddyhoran reviewed (approved):
LGTM

jorgecarleitao commented:
Thanks all for the suggestions. Really good stuff.

I have incorporated all the changes:

* the change of the requirement for performance was re-written as suggested by @alamb and @Dandandan 
* added a code snipped with the idea from @andygrove
* fixed typos identified by @paddyhoran 

I have used the term `soundness` instead of `correctness` because the former is already defined on an official guide: https://rust-lang.github.io/unsafe-code-guidelines/glossary.html#soundness-of-code--of-a-library and is based on existing literature, while the latter seems to be used only by imageproc.

nevi-me reviewed (approved):
I don't have any additional comments

alamb commented:
I think this PR has enough approvals / 👍  so I merged it in. We can continue to iterate on it in the future but this is a great base to work from. Thank you @jorgecarleitao and everyone else who contributed. A true team effort 👍

timClicks commented:
Thanks all for working hard to keep Arrow as robust as possible
